block-layer/codecs/dag-jose.md

```diff
@@ -0,0 +1,43 @@
+# Specification: DAG-JOSE
+
+**Status: Descriptive - Draft**
+
+JOSE is a stanard for signing and encrypting JSON objects. The various specifications for JOSE can be found in the [IETF datatracker](https://datatracker.ietf.org/wg/jose/documents/). 
+
+DAG JOSE supports the full [IPLD Data Model](../data-model-layer/data-model.md).
+
+## Format
+
+The are two main ways to represent a JOSE node. As a JWS ([json web signature](https://datatracker.ietf.org/doc/rfc7515/?include_text=1)) and JWE ([json web encryption](https://datatracker.ietf.org/doc/rfc7516/?include_text=1)). These two formats  acts as the primitives in JOSE and can be used to create JWT and JWM objects. This specification describes how to encode JWS and JWE as an IPLD format.
+
+### Serialization
+
+Both JWS and JWE supports different serialization formats: `Compact Serialization`, `Flattened JSON Serialization`, and `General JSON Serialization`. The first two are more concise, but they only allow for one recipient. Therefore DAG JOSE always uses the General Serialization format for maximum compatibility and minimum ambiguity. 
+
+#### Ordering
+
+Codec implementors **MUST** use the specified order of JOSE properties to ensure hashes consistently match for the same block data. Since JWS and JWE have a strict set of properties this is straight forward.
+
+##### JWS
+
+The top level object has two properties which should have the order: `payload` then `signatures`. The `signatures` property contains an array of signature elements. Within each of these elements there are three properties which should have the order: `protected`, `header`, then `signature`. Important to note here is that `protected` and `header` may be absent.
+
+The content of the `payload`, `signature`, and `protected` properties are `base64url` encoded and therefore does not need any sorting. In contrast, the `header` property contains an unencoded JSON object and should sort object keys by their (UTF-8) encoded representation, i.e. with byte comparisons.
+
+Finally all whitespace should be stripped. This produces the most compact and consistent representation which will ensure that two codecs producing the same data end up with matching block hashes.
+
+##### Deserializing JWS
+
+In it's serialized format a JWS `payload` is encoded using `base64url`. However, this payload may contain IPLD links. Therefore the decoded content of `payload` uses the same approach as [DAG-JSON](./dag-json.md) to support *Bytes Kind* and *Link Kind*. When the DAG-JOSE codec decodes a JWS it should also decode the payload.
```

Comment:
This part isn’t entirely clear to me.

* Does this add `Link` and `Binary` type support for the payload property? If so, there needs to be a lot more specified, like how the signing works (I assume you’d sign the binary representation) and how encryption works for links (is it a `raw` link to encrypted data or is the link itself encrypted?

* Should the codec decode all `base64url` encoded properties to their property Binary representation?

Reply:
From what I understand, these formats allow for encoding arbitrary data, not just JSON formatted data, so we we restricting the allowable payloads to be just JSON, and specifically DAG-JSON? If so, that should be made much clearer.

Reply:
Thanks @mikeal and @rvagg!
Added more description of how they payload works.

> Should the codec decode all base64url encoded properties to their property Binary representation?

I don't think it should since most JOSE tools already know how to interpret the base64url encoded data. Also, signature verification happens over `base64url(header).base64url(payload)` so if we converted to binary representation we would need to convert back before we can verify the signature.

> From what I understand, these formats allow for encoding arbitrary data, not just JSON formatted data, so we we restricting the allowable payloads to be just JSON, and specifically DAG-JSON? If so, that should be made much clearer.

Very good point! We should support both so added some clarification about this.

Reply:
How would we distinguish between JSON and DAG-JSON if they are just arbitrary payload values?

I’m somewhat incline to say that, anything other than inline binary, must be a link and the link is what gets hashed. Not sure what you’d need to do for encryption 🤷‍♂️

Reply:
> How would we distinguish between JSON and DAG-JSON if they are just arbitrary payload values?

Just try to parse it as JSON.

> I’m somewhat inclined to say that anything other than inline binary must be a link and the link is what gets hashed. Not sure what you’d need to do for encryption 🤷‍♂️

Don't think this is a good idea since it would deviate from the JOSE specification, especially for encrypted data. It would also make all existing JOSE objects incompatible with this codec. The idea here is that all existing JOSE objects should be possible to put into ipld.

Reply:
> Just assume DAG-JSON if there is either Binary or Link types in the JSON structure. Basically just always run the object though the DAG-JSON decoder.

It's not quite that simple unfortunately, JSON imposes other strictness requirements above standard JSON you'd also need to check. There's also the weird, but possible case of someone encoding something with JSON that just happens to have elements that could be decoded as DAG-JSON Binary and/or Link. You should probably just say that it's strictly DAG-JSON data and be done with that.

Reply:
@rvagg It's not really possible to strictly use DAG-JSON because it imposes a specific ordering of keys. A JOSE object might not use the same ordering and the content would differ. 
However ordering only really can be enforced when encoding. So I would be fine with always assuming DAG-JOSE when decoding if no reordering of keys is happening.

Reply:
not for the whole doc, but you could specify that the encoded payload should be processed as DAG-JSON, right?

Reply:
Yes, however if they payload object is not sorted in the way DAG-JSON prescribes, we can't change that since it would make the signature invalid if you want to verify it.

Reply:
Oh, the format of data that gets encrypted has its own sorting rules for JSON? I thought it was arbitrary data that gets encoded into the `payload` field?

Reply:
Or are we writing this in a way that handles _any_ JOSE document that happens to have JSON in its payload? If that's the case, maybe we need two codecs, `jose` and `dag-jose`, as we do for `json` and `dag-json` (and `cbor` and `dag-cbor`). But that might depend (for now) on what _your_ use-case is—do you have enough control over JOSE content that you could encode as strict DAG-JSON? If so, maybe we don't need a generic `jose` codec (for now) and just lock this in as strictly DAG-JSON payloads only?

Reply:
> Oh, the format of data that gets encrypted has its own sorting rules for JSON?

No it doesn't and for encryption (JWE) it doesn't matter. However for signed objects (JWS) the library that performs the signature might have a specific way of ordering the payload. So if we tamper with this sorting we will get in trouble. 

I think we can fit everything into this codec without much of an issue actually. I might need to clarify some things still though :)

Reply:
> How would we distinguish between JSON and DAG-JSON if they are just arbitrary payload values?

You can use the `cty` property to address this.
https://tools.ietf.org/html/rfc7516#section-4.1.12

Reply:
for reference, this is the relevant section in the [JWS spec](https://www.rfc-editor.org/rfc/rfc7515.html#section-4.1.10)

have we figured out the right way to represent multicodecs as mime types? i know we’ve talked about the representation in the other direction a lot. I bet @jonnycrunch has thought about this more than I have.